More fixes to forward header parsing

richdougherty opened the pull request:
Fixes #5299.

Parsing for `Forward` and `X-Forwarded-*` headers no longer throws exceptions if an invalid header is found. Now a debug log message is thrown instead. This reverses the stricter errors introduced in #5266.

Additional changes were made:
- Don't bother parsing the supplied IP address headers if we don't trust the connection that supplied it.
- Support values in Forward headers surrounded with double quotes. This is needed to support IPv6.
- Support IPv6 values in X-Forwarded-For headers even when they don't have surrounding square brackets. This format is commonly supplied by proxies.

richdougherty commented on framework/src/play-server/src/main/scala/play/core/server/common/ForwardedHeaderHandler.scala:
```diff
@@ -71,42 +70,30 @@ private[server] class ForwardedHeaderHandler(configuration: ForwardedHeaderHandl
-          // There is a forwarded header from 'prev', so lets process it and get the
-          // address.
+      // Check if there's a forwarded header for us to scan.
+      if (headerEntries.hasNext) {
```
Check `hasNext` first, because we don't need to do `isTrustedProxy` check unless there's another header entry to process.

richdougherty commented on framework/src/play-server/src/main/scala/play/core/server/common/ForwardedHeaderHandler.scala:
```diff
@@ -71,42 +70,30 @@ private[server] class ForwardedHeaderHandler(configuration: ForwardedHeaderHandl
-                  |remove proxy from $TrustedProxiesConfigPath config entry""".stripMargin)
+          configuration.parseEntry(entry) match {
+            case Left(error) =>
+              ForwardedHeaderHandler.logger.debug(
```
Log errors on debug rather than throwing exceptions.

richdougherty commented on framework/src/play-server/src/main/scala/play/core/server/common/ForwardedHeaderHandler.scala:
```diff
@@ -120,48 +107,101 @@ private[server] class ForwardedHeaderHandler(configuration: ForwardedHeaderHandl
+     * objects. Each object a pair with an optional unparsed address and an
+     * optional unparsed protocol. Further parsing may happen later, see `remoteConnection`.
+     */
+    def forwardedHeaders(headers: Headers): Seq[ForwardedEntry] = version match {
```
I made `forwardedHeaders` a standard method rather than an anonymous function. This is just for readability and is unrelated to the fix.

richdougherty commented on framework/src/play-server/src/main/scala/play/core/server/common/ForwardedHeaderHandler.scala:
```diff
@@ -120,48 +107,101 @@ private[server] class ForwardedHeaderHandler(configuration: ForwardedHeaderHandl
+            if (v.length >= 2 && v.charAt(0) == '"' && v.charAt(v.length - 1) == '"') {
+              v.substring(1, v.length - 1)
+            } else v
+          }
```
This change to remove quotes is the main logic change here. It allows us to read values surrounded by quotes. The parsing logic is not exactly the same as the RFC, and it potentially makes us vulnerable to funny header injection tricks, but this is an existing issue and I didn't want to spend days writing an RFC-compliant field parser in this PR!

DorianGrey replied:
It might be a theoretical issue, but ... `splitted` might contain less than two entries in case the provided header is malformed, which would cause an `ArrayIndexOutOfBoundsException` here. You might want to check this for logging these entries, or just for skipping them, like:
```
.map(_.split("=", 2))
.filter(_.length == 2)
.map(splitted => { // the rest of the code) })
```
(I'd thought about .collect here, but splitting twice would be rather unlikely.)

richdougherty commented on framework/src/play-server/src/main/scala/play/core/server/common/ForwardedHeaderHandler.scala:
```diff
@@ -120,48 +107,101 @@ private[server] class ForwardedHeaderHandler(configuration: ForwardedHeaderHandl
+          ForwardedEntry(paramMap.get("for"), paramMap.get("proto"))
+        }
+      case Xforwarded =>
+        def h(h: Headers, key: String) = h.getAll(key).flatMap(s => s.split(",\\s*"))
```
Note we are *not* supported quoting with double quotes in the `X-Forwarded-*` headers. We can add this if it's used in practice.

DorianGrey replied:
They might be present (optional), but I can tell from both Apache and Nginx that they don't use them.

richdougherty replied:
OK I'll support more quoting styles.

richdougherty commented on framework/src/play-server/src/main/scala/play/core/server/common/NodeIdentifierParser.scala:
```diff
@@ -34,11 +30,21 @@ private[common] object NodeIdentifierParser extends RegexParsers {
+        case x: Inet4Address => Ip(x)
+        case x: Inet6Address => Ip(x)
+      }
   }
```
Switch parsing based on which type of header we're in. RFC 7239 supports more types of address, X-Forwarded-For is more forgiving about IPv6 addresses without surrounding square brackets.

DorianGrey replied:
Wouldn't this be a slightly better place to check the (potentially optional) existence of double quotes? (See the other comment)

richdougherty commented on framework/src/play-server/src/test/scala/play/core/server/common/ForwardedHeaderHandlerSpec.scala:
```diff
@@ -17,8 +17,68 @@ class ForwardedHeaderHandlerSpec extends Specification {
+      results(7)._1 must_== ForwardedEntry(Some("unknown"), None)
+      results(7)._2 must beLeft
+      results(7)._3 must beNone
+    }
```
Do lots of tests here to validate the internal parsing logic.

richdougherty commented on framework/src/play-server/src/test/scala/play/core/server/common/ForwardedHeaderHandlerSpec.scala:
```diff
@@ -117,13 +207,44 @@ class ForwardedHeaderHandlerSpec extends Specification {
+        """
+          |X-Forwarded-For: ::1
+        """.stripMargin)) mustEqual ConnectionInfo(addr("::1"), false)
+    }
```
This test captures the bug reported in #5299.

gmethvin commented:
I'm not too familiar with this code but the changes lgtm.

DorianGrey commented on framework/src/play-server/src/main/scala/play/core/server/common/ForwardedHeaderHandler.scala:
```diff
@@ -120,48 +107,101 @@ private[server] class ForwardedHeaderHandler(configuration: ForwardedHeaderHandl
+          val name = splitted(0).toLowerCase(java.util.Locale.ENGLISH)
+          val value = {
+            // Remove quotes, if any
+            val v = splitted(1)
```
It might be a theoretical issue, but ... `splitted` might contain less than two entries in case the provided header is malformed, which would cause an `ArrayIndexOutOfBoundsException` here. You might want to check this for logging these entries, or just for skipping them, like:
```
.map(_.split("=", 2))
.filter(_.length == 2)
.map(splitted => { // the rest of the code) })
```
(I'd thought about .collect here, but splitting twice would be rather unlikely.)

richdougherty replied:
OK I've made this more forgiving and added some tests.

DorianGrey commented on framework/src/play-server/src/main/scala/play/core/server/common/ForwardedHeaderHandler.scala:
```diff
@@ -120,48 +107,101 @@ private[server] class ForwardedHeaderHandler(configuration: ForwardedHeaderHandl
+          val value = {
+            // Remove quotes, if any
+            val v = splitted(1)
+            if (v.length >= 2 && v.charAt(0) == '"' && v.charAt(v.length - 1) == '"') {
```
Another (minor) thing, just for completeness: 
>It is important to note that an IPv6 address and any nodename with
   node-port specified MUST be quoted, since ":" is not an allowed
   character in "token".

See the last paragraph of [this](https://tools.ietf.org/html/rfc7239#section-6) .

I.e.: Any IPv6-address in the value of a `Forwarded` header has to be quoted, and the header should be considered invalid otherwise. Should not happen that often, indeed.

DorianGrey replied:
Just another (minor) issue:
> Note that as ":" and "[]" are not valid characters in "token", IPv6
   addresses are written as "quoted-string".

See [here](https://tools.ietf.org/html/rfc7239#section-4), below the examples.
I.e.: Within a `Forwarded` header, in case of an IPv6-address inside, the quote are **not optional**.

richdougherty replied:
Unfortunately we're not equipped to enforce quoting rules properly unless we write a proper HTTP header field parser. The parser would need to properly follow the grammar and work from left to right. It couldn't use `split` at all, since a split shouldn't happen inside quotes.

I think that writing a full header parser is too much work for this fix so I'm just refining the existing splitting logic to detect quotes and remove them if they're present. This makes the Play logic more forgiving than in the RFC and also means we're still vulnerable to HTTP header smuggling issues, but those issues were already present…

DorianGrey commented:
Just added two (minor) thing that came to my mind which shouldn't be that critical.

typesafehub-validator commented:
Hi @cezary-draus,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.typesafe.com/contribute/cla](http://www.typesafe.com/contribute/cla)

richdougherty commented:
Thanks a lot for the review!

I've updated this now based on review comments. I've also added a fix that's equivalent to #5308, although I didn't use that commit as it still needs CLA validation.

This PR should now be ready to be merged.

richdougherty commented:
@gmethvin, would you be able to merge this for me? :)

gmethvin commented:
:+1:

richdougherty commented:
Thanks!

cezary-draus commented:
Hi!
I sent form accepting CLA, should I send paper scan of CLA somewhere?

2015-11-26 22:59 GMT+01:00 Typesafe CLA Validator <notifications@github.com>
:

> Hi @cezary-draus <https://github.com/cezary-draus>,
>
> Thank you for your contribution! We really value the time you've taken to
> put this together.
>
> Before we proceed with reviewing this pull request, please sign the
> Typesafe Contributors License Agreement:
>
> http://www.typesafe.com/contribute/cla
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/playframework/playframework/pull/5306#issuecomment-159997982>
> .
>



-- 
Pozdrawiam,
Cezary Draus

richdougherty commented:
@cezary-draus, thanks for filling out the CLA form, that's all you need to do. You don't need to send a paper scan.
